src/librustc_typeck/check/demand.rs

```diff
@@ -421,6 +421,9 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
                                 }
                             }
                         }
+                        // If this expression had a clone call, when suggesting borrowing, we
+                        // want to suggest removing it
+                        let sugg_expr = sugg_expr.trim_end_matches(".clone()");
```

Comment:
`Clone` is a lang item; it seems to me more correct to check for that instead, otherwise you could mistakenly remove `.clone()` when it refers to some other method.

Reply:
I'm not sure how we could identify wether it is actually `Clone::clone` here. We have access to the `expr` and `expected_ty`. I believe that at this point the only two things I can figure out are that it was a `ExprKind::MethodCall` of `clone()` (which I also get with the string comparison) and that both the expected and evaluated types are `std::string::String`:

```
DEBUG 2019-05-25T18:48:12Z: rustc_typeck::check::demand: node MethodCall(PathSegment { ident: clone#0, hir_id: Some(HirId { owner: DefIndex(12), local_id: 12 }), res: Some(Err), args: None, infer_types: true }, src/test/ui/issues/issue-61106.rs:3:11: 3:16, [expr(HirId { owner: DefIndex(12), local_id: 14 }: x)])
DEBUG 2019-05-25T18:48:12Z: rustc_typeck::check::demand: ty std::string::String std::string::String
```

Would #60966 help with the reliability here?

In the large scheme of things, this is just a nice to have that could be covered in clippy in a general manner: identify unecessary `clone`s that get immediately borrowed, but I'd love it if we can add it to `rustc` directly.

Reply:
@eddyb @oli-obk would you know of the top of your head of a way to check the above to verify whether the `ExprKind::MethodCall` corresponds to `Clone::clone`?

Reply:
I believe the result of resolving the method call is in `type_dependent_def`.

Reply:
Done!